File: 3420.count-non-decreasing-subarrays-after-k-operations.java
class Solution {
    public long countNonDecreasingSubarrays(int[] nums, int k) {
        int n = nums.length;
        long count = 0;
        for (int start = 0; start < n; start++) {
            int operations = 0;
            for (int end = start; end < n; end++) {
                if (end > start && nums[end] < nums[end - 1]) {
                    operations += nums[end - 1] - nums[end];
                }
                if (operations <= k) {
                    count++;
                } else {
                    break; // No further subarray starting from 'start' can be valid
                }
            }
        }
        return count;
    }
}

File: 3420.count-non-decreasing-subarrays-after-k-operations.go
#
# @lc app=leetcode id=3420 lang=golang
#
# [3420] Count Non-Decreasing Subarrays After K Operations
#

# @lc code=start
func countNonDecreasingSubarrays(nums []int, k int) int64 {
    n := len(nums)
    result := int64(0)
    left := 0
    currentOperations := 0
    for right := 1; right < n; right++ {
        if nums[right] < nums[right-1] {
            currentOperations += nums[right-1] - nums[right]
        }
        while currentOperations > k && left < right {
            if nums[left+1] < nums[left] {
                currentOperations -= nums[left] - nums[left+1]
            }
            left++
        }
        result += int64(right - left)
    }
    result += int64(n) // each single element is a valid subarray
    return result
}
# @lc code=end